src/p_udmf.c

```diff
@@ -0,0 +1,937 @@
+    {
+        // [FG] build nodes with NanoBSP
+        BSP_BuildNodes();
+    }
+    else
+    {
+        // support all ZDoom extended node formats
+        P_LoadNodes_ZDoom(znodes_num, nodeformat);
+    }
+
+    return nodeformat;
+}
+
+boolean UDMF_LoadBlockMap(int blockmap_num)
+{
+    long count;
+    boolean ret = true;
+
+    //!
+    // @category mod
+    //
+    // Forces a (re-)building of the BLOCKMAP lumps for loaded maps.
+    //
+
+    // [FG] always rebuild too short blockmaps
+    if (M_CheckParm("-blockmap")
+        || (count = W_LumpLengthWithName(blockmap_num, "BLOCKMAP") / 2)
+               >= 0x10000
+        || count < 4)
+    {
+        P_CreateBlockMap();
+    }
+    else
+    {
+        long i;
+        short *wadblockmaplump = W_CacheLumpNum(blockmap_num, PU_LEVEL);
+        blockmaplump = Z_Malloc(sizeof(*blockmaplump) * count, PU_LEVEL, 0);
+
+        // killough 3/1/98: Expand wad blockmap into larger internal one,
+        // by treating all offsets except -1 as unsigned and zero-extending
+        // them. This potentially doubles the size of blockmaps allowed,
+        // because Doom originally considered the offsets as always signed.
+
+        blockmaplump[0] = SHORT(wadblockmaplump[0]);
+        blockmaplump[1] = SHORT(wadblockmaplump[1]);
+        blockmaplump[2] = (long)(SHORT(wadblockmaplump[2]))&FRACMASK;
+        blockmaplump[3] = (long)(SHORT(wadblockmaplump[3]))&FRACMASK;
+
+        for (i = 4; i < count; i++)
+        {
+            short t = SHORT(wadblockmaplump[i]); // killough 3/1/98
+            blockmaplump[i] = t == -1 ? -1l : (long)t & FRACMASK;
+        }
+
+        Z_Free(wadblockmaplump);
+
+        bmaporgx = blockmaplump[0] << FRACBITS;
+        bmaporgy = blockmaplump[1] << FRACBITS;
+        bmapwidth = blockmaplump[2];
+        bmapheight = blockmaplump[3];
+
+        ret = false;
+
+        P_SetSkipBlockStart();
+    }
+
+    // clear out mobj chains
+    blocklinks_size = sizeof(*blocklinks) * bmapwidth * bmapheight;
+    blocklinks = Z_Malloc(blocklinks_size, PU_LEVEL, 0);
+    memset(blocklinks, 0, blocklinks_size);
+    blockmap = blockmaplump + 4;
+
+    return ret;
+}
+
+boolean UDMF_LoadReject(int reject_num, int totallines)
+{
+    // Calculate the size that the REJECT lump *should* be.
+    int minlength = (numsectors * numsectors + 7) / 8;
+    int lumplen = W_LumpLengthWithName(reject_num, "REJECT");
+    boolean ret;
+
+    // If the lump meets the minimum length, it can be loaded directly.
+    // Otherwise, we need to allocate a buffer of the correct size
+    // and pad it with appropriate data.
+
+    if (lumplen >= minlength)
+    {
+        rejectmatrix = W_CacheLumpNum(reject_num, PU_LEVEL);
+        ret = false;
+    }
+    else
+    {
+        unsigned int padvalue = 0x00;
+
+        rejectmatrix = Z_Malloc(minlength, PU_LEVEL, (void **)&rejectmatrix);
+        W_ReadLump(reject_num, rejectmatrix);
+
+        if (M_CheckParm("-reject_pad_with_ff"))
```

Comment:
Do we need to copy this function (and `LoadBlockMap`) from `p_setup.c`? I don't think anyone will convert an old vanilla map to UDMF and maintain the old REJECT lump in order to be compatible with vanilla demos.

Reply:
> What does this fix exactly? The control flow shouldn't even reach this area if there's no REJECT lump.

Both [UDMF_test.zip](https://github.com/user-attachments/files/22581351/UDMF_test.zip) and [vandacts_udmf.zip](https://github.com/user-attachments/files/22583131/vandacts_udmf.zip) crash for me without this fix. It tries to call `W_ReadLump` with lump num -1. I think we don't need `W_ReadLump` call at all, it's from Choco.